gtfs-validator/main/src/main/java/org/mobilitydata/gtfsvalidator/validator/StopsTooCloseValidator.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.mobilitydata.gtfsvalidator.validator;
+
+import static org.locationtech.spatial4j.context.SpatialContext.GEO;
+
+import com.google.common.collect.Multimaps;
+import java.lang.Double;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import org.locationtech.spatial4j.distance.DistanceCalculator;
+import org.locationtech.spatial4j.distance.DistanceUtils;
+import org.locationtech.spatial4j.shape.Point;
+import org.locationtech.spatial4j.shape.ShapeFactory;
+import org.mobilitydata.gtfsvalidator.annotation.GtfsValidator;
+import org.mobilitydata.gtfsvalidator.annotation.Inject;
+import org.mobilitydata.gtfsvalidator.notice.NoticeContainer;
+import org.mobilitydata.gtfsvalidator.notice.StopsTooCloseNotice;
+import org.mobilitydata.gtfsvalidator.table.GtfsLocationType;
+import org.mobilitydata.gtfsvalidator.table.GtfsStop;
+import org.mobilitydata.gtfsvalidator.table.GtfsStopTableContainer;
+
+/**
+ * Validates that two stops are not too close to each other.
+ *
+ * <p>Generated notice: {@link StopsTooCloseNotice}.
+ */
+@GtfsValidator
+public class StopsTooCloseValidator extends FileValidator {
+  private static final int KILOMETER_TO_METER_CONVERSION_FACTOR = 1000;
+  /**
+  * At equator 1 degree = 110.567 kilometers.
+  * 0.00005 degrees = (110.567km * 1000) * 0.00005 = 5.52835 m
+  * At poles 1 degree = 111.699 kilometers.
+  * 0.00005 degrees = (111.699km * 1000) * 0.00005 = 5.58495 m
+  *
+  * So the latitude buffer is 0.00005 degrees to check stops within at least 5.52835 m 
+  */
+  private static final double LAT_BUFFER = 0.00005;
```

Comment:
Could this value be derived from DIST_BUFFER_METERS (possibly using
https://locationtech.github.io/spatial4j/apidocs/org/locationtech/spatial4j/distance/DistanceUtils.html )?

Reply:
They only have DistanceUtils.calcLonDegreesAtLat(), but not the opposite to calculate Lat Degrees. However, this takes the distance in degrees and not meters. Then there is dist2Degrees(double dist, double radius) but requires the radius of the earth which varies as well. So it doesn't appear there is a straight forward calculation.

Reply:
I think you could use a spherical earth model for this calculation, (using https://locationtech.github.io/spatial4j/apidocs/org/locationtech/spatial4j/distance/DistanceUtils.html#EARTH_EQUATORIAL_RADIUS_KM )

Reply:
I think it makes more sense to have the buffer as latitude instead of longitutde: 

Degrees of latitude are parallel so the distance between each degree remains almost constant but since degrees of longitude are farthest apart at the equator and converge at the poles, their distance varies greatly.

A degree of longitude is widest at the equator at 69.172 miles (111.321) and gradually shrinks to zero at the poles.

Therefore our buffer would have to be really large to account for the fact it shrinks to 0 at the poles

Reply:
I think that dist2Degrees should produce a correct value.

When you try to figure out the distance between points on a sphere based on an angle, you need to to take the angle in the plane containing the sphere center and the two endpoints (not only latitude and longitude).

Using spherical approximation would provide the correct result with maximal 0.3% error (https://gis.stackexchange.com/questions/25494/how-accurate-is-approximating-the-earth-as-a-sphere)

Reply:
It will produce the correct value, it's just that 1 degree = 0 km at the poles, so to check you would technically need a longitude_buffer =  ∞, rather than a latitude one = 0.00005

Reply:
I could instead create my own private variable METERS_TO_LATITUDE_DEGREES_CONVERSION_FACTOR =  110.567km and then multiply it by the DIST_BUFFER_METERS, instead of using these functions? If that's ok?

Reply:
1 degree difference in the pole >> 0 km (as long as the vertex from which you calculate the rays is the center of earth).
1 latitude degree difference in the pole is 0 km.

I am proposing using a LAT_BUFFER = dist2Degrees(DIST_BUFFER_METERS, EARTH_EQUATORIAL_RADIUS_KM  * KILOMETER_TO_METER_CONVERSION_FACTOR).
Can you please help me understand why using 0.00005 would be better?

Reply:
Ok yes I understand now. I just realised I could use this function to convert. I initially was just looking at the DistanceUtils.calcLonDegreesAtLat() function and saw that they didn't have a calcLatDegreesAtLon() but using the EARTH_EQUATORIAL_RADIUS_KM we can calculate the conversion at the radius. I didn't quite understand how to use the function. But I have now implemented it as so.